File: swiftbar_plugin/transit_eta/ojp_http.py
"""Shared bits for talking to the OJP endpoint: auth, timestamps, errors.

See the module docstring in ojp_client.py for the caveat on why
OJP_ENDPOINT needs to be double-checked against opentransportdata.swiss's
live docs. The auth scheme (Authorization: Bearer <key>, plus a
non-empty User-Agent) is confirmed against opentransportdata.swiss's own
docs -- their gateway 403s requests with no User-Agent set, which is
Python's urllib default.
"""

import urllib.error
from datetime import datetime, timezone
from typing import Dict

OJP_ENDPOINT = "https://api.opentransportdata.swiss/ojp20"
REQUEST_TIMEOUT_SECONDS = 15
USER_AGENT = "transit-eta-swiftbar/1.0"


class OjpError(RuntimeError):
    """Raised for network/auth/parsing failures talking to the OJP API."""


def auth_headers(api_key: str) -> Dict[str, str]:
    return {
        "Authorization": f"Bearer {api_key}",
        "User-Agent": USER_AGENT,
    }


def iso_now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def error_from_http_error(exc: "urllib.error.HTTPError") -> OjpError:
    # exc.reason is just the status phrase (e.g. "Forbidden"); the body
    # usually carries the actual reason (invalid/expired key, not
    # subscribed to this product, quota exceeded, ...), so surface it.
    try:
        detail = exc.read().decode("utf-8", errors="replace").strip()
    except Exception:
        detail = ""
    detail = detail[:300] if detail else exc.reason
    return OjpError(f"HTTP {exc.code} from OJP API: {detail}")


def xml_escape(text: str) -> str:
    return (
        text.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
    )

File: TransitETAApp/TransitETAApp/OJPHTTP.swift
import Foundation

let ojpEndpoint = URL(string: "https://api.opentransportdata.swiss/ojp20")!
let requestTimeoutSeconds: TimeInterval = 15
let userAgent = "transit-eta-swiftui/1.0"

struct OjpError: Error, LocalizedError, Equatable {
    let message: String
    var errorDescription: String? { message }
}

func authHeaders(apiKey: String) -> [String: String] {
    ["Authorization": "Bearer \(apiKey)", "User-Agent": userAgent]
}

func isoNow() -> String {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
    formatter.timeZone = TimeZone(identifier: "UTC")
    formatter.locale = Locale(identifier: "en_US_POSIX")
    return formatter.string(from: Date())
}

func xmlEscape(_ text: String) -> String {
    text.replacingOccurrences(of: "&", with: "&amp;")
        .replacingOccurrences(of: "<", with: "&lt;")
        .replacingOccurrences(of: ">", with: "&gt;")
        .replacingOccurrences(of: "\"", with: "&quot;")
}

// exc.reason from urllib is just the status phrase; the body usually
// carries the actual reason (invalid/expired key, quota exceeded, ...),
// so surface it -- mirrors ojp_http.py's error_from_http_error.
func ojpError(fromHTTPResponse response: HTTPURLResponse, body: Data) -> OjpError {
    var detail = String(data: body, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    if detail.isEmpty {
        detail = HTTPURLResponse.localizedString(forStatusCode: response.statusCode)
    } else if detail.count > 300 {
        detail = String(detail.prefix(300))
    }
    return OjpError(message: "HTTP \(response.statusCode) from OJP API: \(detail)")
}
